### ISSUE TITLE
Yaml deserializer conflict with `readf`

### ISSUE THREAD

**Issue by jeremiah-corrado** [chapel-lang/chapel#22787]
### Summary of problem

The `%t` format specifier (soon to be replaced by `%?` in https://github.com/chapel-lang/chapel/pull/22766), can be used to invoke a `fileReader`'s deserializer to parse some input.

This is currently not working with the Yaml deserializer, because it uses seeking under the hood, but the `readf` implementation marks the file position making seeking impossible.

### Steps to reproduce

To compile the following program, `libyaml` must be installed and its path information must be supplied to the chapel compiler (along with a couple of other flags related to CTypes deprecations — at the time of writing). Example:
```bash
chpl --ccflags='-I/usr/local/include' --ldflags='-L/usr/local/lib -lyaml' -scFileTypeHasPointer=false -scPtrToLogicalValue=true program.chpl
```

```chapel
use IO, YAML;

record R {
  var x: int;
}

var f = openTempFile(),
    r: R;

f.writer(serializer = new YamlSerializer()).write(new R(1));
f.reader(deserializer = new YamlDeserializer()).readf("%?", r);

writeln(r);
```

When run, it produces the following error:
```
uncaught SystemError: Invalid argument: reset not supported for marked channel
  $CHPL_HOME/modules/standard/IO.chpl:4062: thrown here
  program.chpl:4: uncaught here
```

**Comment by mppf** [comment]
I'd expect many/most of our serializer and deserializers to work with Sockets or Pipes or other kinds of "files" that don't allow seeking. It would not bother me if that is impossible for certain specialized formats, but I am inclined to agree with the "bug" designation here and say that the Yaml support should work on Sockets/Pipes/etc.

**Comment by jeremiah-corrado** [comment]
I agree. Unfortunately this bug originates from the fact that `YamlDeserializer` is built on top of libyaml's parser which requires its own direct access to the given `FILE*`. As a result, the `fileReader` needs to sort of follow the parser around the file to read whatever it finds. Example:
https://github.com/chapel-lang/chapel/blob/6e051492dded4e07a892697412a36e692eafd540/modules/packages/YAML.chpl#L634-L639
*here, `yamlReader` is an alias for `fileReader(YamlDeserializer, ?)`

Ultimately, to solve this problem, the `fileReader` will likely have to read chunks of data into an intermediate buffer which is provided to the libyaml parser in the form of a `FILE*`, which introduces more complexity than seemed worthwhile for the prototype.

**Comment by bradcray** [comment]
Here's a very naive question (along a number of axes):  Would it be possible to have the YAML deserializer read the entire file/pipe/source into memory and then do the seeking and deserializer within that memory buffer?  Or is it necessary to do it in the original FILE* for technical, semantic, or IO framework reasons?

**Comment by jeremiah-corrado** [comment]
The `libyaml` parser either needs direct access to a `FILE*` or a `char *`. This is a restriction from the library's API. I chose not to read the entire contents of the `fileReader` into a string (or another `FILE*`) and hand that off to the parser for a couple of reasons:
- there could be a `fileWriter` or some other mechanism modifying the actual file simultaneously. This could cause the following problem:
   ```chapel
   var f = open("config.yaml"),
       r = f.reader(deserializer = new YamlDeserializer()),
       w = f.writer(serializer = new YamlSerializer());

   var x = r.read(someRecordType); // consumes the entire file under the hood to populate the parser's buffer
   w.write(new otherRecordType(blah)); // add something else to the actual file
   
   // any subsequent reads don't include the `otherRecordType`
   r.read(/* something else */);
   ```
- could cause some memory overhead issues when parsing a large file (probably unlikely for YAML, but possible)
